src/storage.py

```diff
@@ -0,0 +1,235 @@
+            # if no errors while removing OSD, clean stale osd records.
+            self._clean_stale_osd_data()
+        except CalledProcessError as e:
+            if force:
+                # If forced removal was done, clean stale osd records.
+                self._clean_stale_osd_data()
+            raise e
+
+    def _save_osd_data(self, disk_name: str, db_name: str = None):
+        """Save OSD data using juju storage names."""
+        disk_path = self.juju_storage_get(storage_id=disk_name, attribute="location")
+
+        for osd in microceph.list_disk_cmd()["ConfiguredDisks"]:
+            # get block device info using /dev/disk-by-id and lsblk.
+            local_device = microceph._get_disk_info(osd["path"])
+
+            # OSD not configured on current unit.
+            if not local_device:
+                continue
+
+            # e.g. check 'vdd' in '/dev/vdd'
+            if local_device["name"] in disk_path:
+                logger.debug(f"Added OSD {osd['osd']} with Disk {disk_name}, DB {db_name}")
+                self._stored.osd_data[osd["osd"]] = {
+                    "disk_by_id": osd["path"],  # /dev/disk-by-id/ for OSD device.
+                    "disk": disk_name,  # storage name for OSD device.
+                    "db": db_name,  # storage name for DB device.
+                }
+
+    def _get_osd_id(self, name: str):
+        """Fetch the OSD number of consuming OSD, None is not used as OSD."""
+        # storage name is of the form db/3 or osd-standalone/2 etc.
+        directive = name.split("/")[0]
+
+        if directive == self.standalone:
+            directive = "disk"
+
+        logger.debug(self._stored.osd_data)
+        logger.debug(f"Searching for disk {name}")
+
+        for k, v in dict(self._stored.osd_data).items():
+            # if value is not None.
+            if v and v[directive] == name:
+                return k  # key is the stored osd number.
+        return None
+
+    def _clean_stale_osd_data(self):
+        """Compare with disk list and remove stale entries."""
+        osds = [osd["osd"] for osd in microceph.list_disk_cmd()["ConfiguredDisks"]]
+
+        for osd_num in dict(self._stored.osd_data).keys():
+            if osd_num not in osds:
+                val = self._stored.osd_data[osd_num]
+                self._stored.osd_data[osd_num] = None
+                logger.debug(f"Popped {val}")
+
+    # NOTE(utkarshbhatthere): 'storage-get' sometimes fires before
+    # requested information is available.
+    @retry(wait=wait_fixed(5), stop=stop_after_attempt(10))
+    def juju_storage_get(self, storage_id=None, attribute=None):
+        """Get storage attributes."""
+        _args = ["storage-get", "--format=json"]
+        if storage_id:
+            _args.extend(("-s", storage_id))
+        if attribute:
+            _args.append(attribute)
+        try:
+            return json.loads(self._run(_args))
+        except ValueError:
```

Comment:
Nit: maybe some logging.debug() so we don't swallow errors silently

Reply:
ack